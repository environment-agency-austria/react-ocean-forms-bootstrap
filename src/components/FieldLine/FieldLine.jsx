/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { Label, Col, InputGroup } from 'reactstrap';
import { FormText } from 'react-ocean-forms';
import { fieldMetaShape, fieldShape } from 'react-ocean-forms-legacy';

import { FieldRow } from './FieldRow';
import { RequiredMarker } from './RequiredMarker';
import { InvalidAlert } from './InvalidAlert';
import { ValidatingSpinner } from './ValidatingSpinner';
import { FieldLineAddon } from './FieldLineAddon';
import { InfoAddonButton } from './InfoAddonButton';
import { InfoAlert } from './InfoAlert';
import FieldError from './FieldError';

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
class FieldLine extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      infoVisible: false,
    };

    this.toggleInfo = this.toggleInfo.bind(this);
  }

  /**
   * Toggles the visibility of the info alert
   */
  toggleInfo() {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  render() {
    const {
      field,
      meta,
      label,
      children,
      validators,
      prefix,
      suffix,
      info,
      labelSize,
      inputSize,
      labelClass,
    } = this.props;

    const { infoVisible } = this.state;

    return (
      <FieldRow meta={meta}>
        <Label sm={labelSize} for={field.id} className={labelClass}>
          <InvalidAlert valid={meta.valid} />
          <ValidatingSpinner isValidating={meta.isValidating} />
          <FormText text={label} />
          <RequiredMarker meta={meta} validators={validators} />
        </Label>
        <Col sm={inputSize}>
          <InputGroup>
            <FieldLineAddon plaintext={meta.plaintext} type="prepend" content={prefix} />
            {children}
            <FieldLineAddon plaintext={meta.plaintext} type="append" content={suffix} />
            <InfoAddonButton info={info} plaintext={meta.plaintext} onClick={this.toggleInfo} />
            <FieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
            />
          </InputGroup>
          <InfoAlert
            visible={infoVisible}
            info={info}
            plaintext={meta.plaintext}
            onClose={this.toggleInfo}
          />
        </Col>
      </FieldRow>
    );
  }
}

FieldLine.displayName = 'FieldLine';

FieldLine.defaultProps = {
  info: undefined,
  validators: undefined,
  prefix: undefined,
  suffix: undefined,
  labelSize: 3,
  inputSize: 9,
  labelClass: 'text-right',
};

FieldLine.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  validators: PropTypes.arrayOf(PropTypes.func),
  prefix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  suffix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  labelSize: PropTypes.string,
  inputSize: PropTypes.string,
  labelClass: PropTypes.string,
};

export default FieldLine;
