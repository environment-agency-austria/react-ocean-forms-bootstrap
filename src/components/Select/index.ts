export * from './SelectBase';
export * from './Select';
export * from './CreatableSelect';
