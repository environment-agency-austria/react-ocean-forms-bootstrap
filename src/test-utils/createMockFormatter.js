export default () => jest.fn().mockImplementation(id => id);
