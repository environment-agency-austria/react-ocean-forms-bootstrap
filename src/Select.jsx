/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import ReactSelect from 'react-select';
import { Input as StrapInput } from 'reactstrap';
import { fieldMetaShape, fieldShape } from 'react-ocean-forms';

import FieldLine from './FieldLine';

/**
 * Component for displaying bootstrap
 * form groups with an select input and
 * oForm support
 */
class Select extends React.Component {
  constructor(props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
  }

  /**
   * Manually handle the onChange event of
   * the select because oForms is expecting
   * default input onChange behaviour.
   */
  handleChange(value) {
    this.props.field.onChange({
      target: {
        name: this.props.field.name,
        value,
      },
    });
  }

  /**
   * Manually handle the onBlur event of
   * the select because oForms is expecting
   * default input onBlur behaviour.
   */
  handleBlur() {
    this.props.field.onBlur({
      target: {
        name: this.props.field.name,
      },
    });
  }

  render() {
    const {
      field,
      placeholder,
      options,
      meta,
    } = this.props;

    // Grab some default texts from the internationalized messages
    const selectPlaceholder = meta.stringFormatter(placeholder);
    const selectNoResults = meta.stringFormatter('ojs_select_noresults');
    const selectClearValue = meta.stringFormatter('ojs_select_clearValue');

    // Generate a css class based on the validity of the select element
    let selectClass = 'react-select-control';
    selectClass = (selectClass + (meta.valid ? '' : ' is-invalid')).trim();

    // Check if the current value has a different label than the value
    // with the same key in the options array. Bugfix to change the
    // selected label when the current language changes.
    if (field.value !== undefined && field.value !== null) {
      const selectableValue = options.find(item => item.value === field.value.value);
      if (selectableValue !== undefined && selectableValue.label !== field.value.label) {
        field.value.label = selectableValue.label;
        this.handleChange(field.value);
      }
    }

    // Support for plaintext display
    if (meta.plaintext) {
      let displayValue = '';
      if (field.value !== undefined && field.value !== null) {
        displayValue = field.value.label;
      }

      return (
        <FieldLine {...this.props}>
          <StrapInput {...field} plaintext>{displayValue}</StrapInput>
        </FieldLine>
      );
    }

    return (
      <FieldLine {...this.props}>
        <ReactSelect
          id={field.id}
          // name={field.name} causes react to crash on change if there's a name?!
          value={field.value}
          disabled={field.disabled}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
          options={options}
          placeholder={selectPlaceholder}
          noResultsText={selectNoResults}
          clearValueText={selectClearValue}
          invalid={!meta.valid}
          className={selectClass}
        />
      </FieldLine>
    );
  }
}

Select.defaultProps = {
  info: undefined,
  placeholder: 'ojs_select_placeholder',
};

Select.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  placeholder: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
  })).isRequired,
};

export default Select;
