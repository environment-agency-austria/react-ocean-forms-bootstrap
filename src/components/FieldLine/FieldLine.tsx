/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';

import { FormText, PropsOf } from 'react-ocean-forms';
import { Col, InputGroup, Label } from 'reactstrap';

import { FieldError } from './FieldError';
import { FieldLineAddon } from './FieldLineAddon';
import { FieldRow } from './FieldRow';
import { InfoAddonButton } from './InfoAddonButton';
import { InfoAlert } from './InfoAlert';
import { InvalidAlert } from './InvalidAlert';
import { RequiredMarker } from './RequiredMarker';
import { ValidatingSpinner } from './ValidatingSpinner';

import { IFieldLineProps } from './FieldLine.types';

interface IFieldLineState {
  infoVisible: boolean;
}

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
export class FieldLine extends React.Component<IFieldLineProps, IFieldLineState> {
  public static displayName: string = 'FieldLine';

  public static defaultProps = {
    labelSize: '3',
    inputSize: '9',
    labelClass: 'text-right',
  };

  constructor(props: IFieldLineProps) {
    super(props);

    this.state = {
      infoVisible: false,
    };
  }

  /**
   * Toggles the visibility of the info alert
   */
  private toggleInfo = (): void => {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  public render(): JSX.Element {
    const {
      field,
      meta,
      label,
      children,
      prefix,
      suffix,
      info,
      labelSize,
      inputSize,
      labelClass,
    } = this.props;

    const { infoVisible } = this.state;

    return (
      <FieldRow meta={meta}>
        <Label sm={labelSize} for={field.id} className={labelClass}>
          <InvalidAlert valid={meta.valid} />
          <ValidatingSpinner isValidating={meta.isValidating} />
          <FormText text={label} />
          <RequiredMarker meta={meta} />
        </Label>
        <Col sm={inputSize}>
          <InputGroup>
            <FieldLineAddon plaintext={meta.plaintext} type="prepend" content={prefix} />
            {children}
            <FieldLineAddon plaintext={meta.plaintext} type="append" content={suffix} />
            <InfoAddonButton info={info} plaintext={meta.plaintext} onClick={this.toggleInfo} />
            <FieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
            />
          </InputGroup>
          <InfoAlert
            visible={infoVisible}
            info={info}
            plaintext={meta.plaintext}
            onClose={this.toggleInfo}
          />
        </Col>
      </FieldRow>
    );
  }
}

export type DefaultizedFieldLineProps = JSX.LibraryManagedAttributes<typeof FieldLine, PropsOf<typeof FieldLine>>;
