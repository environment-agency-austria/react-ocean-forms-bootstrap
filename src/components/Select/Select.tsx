/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import { withField } from 'react-ocean-forms';
import { default as ReactSelect } from 'react-select';
import { Input as StrapInput } from 'reactstrap';

import { FieldLine } from '../FieldLine';
import { ISelectOption, ISelectProps, isSelectOption } from './Select.types';

/**
 * Component for displaying bootstrap
 * form groups with an select input and
 * oForm support
 */
export class BaseSelect extends React.Component<ISelectProps> {
  public static displayName: string = 'Select';

  // tslint:disable-next-line:typedef
  public static defaultProps = {
    multi: false,
    placeholder: 'ojs_select_placeholder',
  };

  /**
   * Manually handle the onChange event of
   * the select because oForms is expecting
   * default input onChange behaviour.
   */
  private handleChange = (value: ISelectOption): void => {
    const { field } = this.props;

    field.onChange({
      target: {
        value,
      },
    });
  }

  /**
   * Manually handle the onBlur event of
   * the select because oForms is expecting
   * default input onBlur behaviour.
   */
  private handleBlur = (): void => {
    const { field } = this.props;

    field.onBlur();
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const {
      field,
      placeholder,
      options,
      multi,
      meta,
    } = this.props;

    // Generate a css class based on the validity of the select element
    let selectClass = 'react-select-control';
    selectClass = (selectClass + (meta.valid ? '' : ' is-invalid')).trim();

    // Check if the current value has a different label than the value
    // with the same key in the options array. Bugfix to change the
    // selected label when the current language changes.
    let fieldValue: ISelectOption | undefined;
    if (isSelectOption(field.value)) {
      fieldValue = field.value;
      // tslint:disable-next-line:no-non-null-assertion
      const selectableValue = options.find(item => item.value === fieldValue!.value);
      if (selectableValue !== undefined && selectableValue.label !== field.value.label) {
        field.value.label = selectableValue.label;
        this.handleChange(field.value);
      }
    }

    // Support for plaintext display
    if (meta.plaintext) {
      let displayValue = '';
      if (isSelectOption(field.value)) {
        displayValue = field.value.label;
      }

      return (
        <FieldLine {...this.props}>
          <StrapInput {...field} value="" plaintext>
            {displayValue}
          </StrapInput>
        </FieldLine>
      );
    }

    const placeholderText = meta.stringFormatter(placeholder);

    return (
      <FieldLine {...this.props}>
        <ReactSelect
          id={field.id}
          value={fieldValue}
          isDisabled={field.disabled}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
          options={options}
          isMulti={multi}
          placeholder={placeholderText}
          className={selectClass}
        />
      </FieldLine>
    );
  }
}

export const Select = withField(BaseSelect);
