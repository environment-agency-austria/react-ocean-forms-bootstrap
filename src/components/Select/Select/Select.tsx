/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { default as ReactSelect } from 'react-select';

import { IPreparedSelectProps, SelectBase, ISelectFieldValue } from '../SelectBase';
import { ISelectProps } from './Select.types';

/**
 * Renders the actual `Select` component with the passed props
 */
const renderSelect = (preparedProps: IPreparedSelectProps): JSX.Element => {
  return (
    <ReactSelect {...preparedProps} />
  );
}

/**
 * Component for displaying bootstrap
 * form groups with an select input and
 * oForm support
 */
export const Select = <TSubmitValue extends unknown = ISelectFieldValue>(props: ISelectProps<TSubmitValue>): JSX.Element => {
  const {
    ...selectBaseProps
  } = props;

  return (
    <SelectBase
      {...selectBaseProps}
      renderSelect={renderSelect}
    />
  );
};
