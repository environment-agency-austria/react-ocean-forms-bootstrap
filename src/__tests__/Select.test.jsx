import React from 'react';
import { shallow } from 'enzyme';
import mockEvent, { KEYCODE } from '../test-utils/enzymeEventUtils';
import createMockFormatter from '../test-utils/createMockFormatter';

import Select from '../Select';
import FieldLine from '../FieldLine';

describe('<Select />', () => {
  const FIELD_ID = 'field0';
  const FIELD_NAME = 'field0';
  const FIELD_LABEL = 'field0';

  const options = [
    { value: 'one', label: 'One' },
    { value: 'two', label: 'Two' },
  ];
  const meta = {
    valid: true,
    error: undefined,
    isValidating: undefined,
    stringFormatter: createMockFormatter(),
  };
  const field = {
    value: '',
    invalid: false,
    id: FIELD_ID,
    name: FIELD_NAME,
    disabled: false,
    onChange: jest.fn(),
    onBlur: jest.fn(),
  };

  const setup = props => shallow((
    <Select
      label={FIELD_LABEL}
      meta={meta}
      field={field}
      options={options}
      {...props}
    />
  ));

  it('should render correctly', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a field line and a bootstrap input', () => {
    const wrapper = setup();

    expect(wrapper.find(FieldLine).exists()).toBeTruthy();
    expect(wrapper.find('Select').exists()).toBeTruthy();
  });

  it('should pass the meta props to the field line', () => {
    const wrapper = setup();

    expect(wrapper.find(FieldLine).prop('meta')).toEqual(meta);
  });

  it('should call field.onChange when the input changes', () => {
    const wrapper = setup();
    const event = { target: { name: FIELD_ID, value: options[0] } };

    // The react-select element doesn't support direct
    // simulate calls of change / blur types on itself.
    // As a workaround we simulate key presses.
    // https://stackoverflow.com/questions/41991077/testing-react-select-component/46201546#46201546
    const selectElement = wrapper.find('Select').dive().find('.Select-control');
    selectElement.simulate('keyDown', mockEvent({ keyCode: KEYCODE.DOWN_ARROW }));
    selectElement.simulate('keyDown', mockEvent({ keyCode: KEYCODE.ENTER }));

    expect(field.onChange).toHaveBeenCalledWith(event);
  });

  it.skip('should call field.onBlur when there is an input blur', () => {
    const wrapper = setup();
    const event = { target: { name: FIELD_ID } };

    const selectElement = wrapper.find('Select').dive().find('.Select-control');
    selectElement.focus();
    selectElement.simulate('keyDown', mockEvent({ keyCode: KEYCODE.DOWN_ARROW }));
    selectElement.simulate('keyDown', mockEvent({ keyCode: KEYCODE.TAB }));

    expect(field.onBlur).toHaveBeenCalledWith(event);
  });
});
