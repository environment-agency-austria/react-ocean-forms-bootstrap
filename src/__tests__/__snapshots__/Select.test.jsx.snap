// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should correctly react to meta.plaintext 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Select
    field={
      Object {
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "target": Object {
                  "name": "mock-field",
                  "value": Object {
                    "label": "One",
                    "value": "one",
                  },
                },
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "value": "",
      }
    }
    label="field0"
    meta={
      Object {
        "error": null,
        "isValidating": false,
        "plaintext": true,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      }
    }
    multi={false}
    options={
      Array [
        Object {
          "label": "One",
          "value": "one",
        },
        Object {
          "label": "Two",
          "value": "two",
        },
      ]
    }
    placeholder="ojs_select_placeholder"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Input
        disabled={false}
        id="mock-field"
        name="mock-field"
        onBlur={[MockFunction]}
        onChange={
          [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "target": Object {
                    "name": "mock-field",
                    "value": Object {
                      "label": "One",
                      "value": "one",
                    },
                  },
                },
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          }
        }
        plaintext={true}
        type="text"
        value=""
      />,
      "field": Object {
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "target": Object {
                  "name": "mock-field",
                  "value": Object {
                    "label": "One",
                    "value": "one",
                  },
                },
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "value": "",
      },
      "info": undefined,
      "label": "field0",
      "meta": Object {
        "error": null,
        "isValidating": false,
        "plaintext": true,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      },
      "multi": false,
      "options": Array [
        Object {
          "label": "One",
          "value": "one",
        },
        Object {
          "label": "Two",
          "value": "two",
        },
      ],
      "placeholder": "ojs_select_placeholder",
      "prefix": undefined,
      "suffix": undefined,
      "validators": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": undefined,
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "target": Object {
                  "name": "mock-field",
                  "value": Object {
                    "label": "One",
                    "value": "one",
                  },
                },
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "plaintext": true,
        "type": "text",
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Input
          disabled={false}
          id="mock-field"
          name="mock-field"
          onBlur={[MockFunction]}
          onChange={
            [MockFunction] {
              "calls": Array [
                Array [
                  Object {
                    "target": Object {
                      "name": "mock-field",
                      "value": Object {
                        "label": "One",
                        "value": "one",
                      },
                    },
                  },
                ],
              ],
              "results": Array [
                Object {
                  "isThrow": false,
                  "value": undefined,
                },
              ],
            }
          }
          plaintext={true}
          type="text"
          value=""
        />,
        "field": Object {
          "disabled": false,
          "id": "mock-field",
          "name": "mock-field",
          "onBlur": [MockFunction],
          "onChange": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "target": Object {
                    "name": "mock-field",
                    "value": Object {
                      "label": "One",
                      "value": "one",
                    },
                  },
                },
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          },
          "value": "",
        },
        "info": undefined,
        "label": "field0",
        "meta": Object {
          "error": null,
          "isValidating": false,
          "plaintext": true,
          "stringFormatter": [MockFunction],
          "touched": false,
          "valid": true,
        },
        "multi": false,
        "options": Array [
          Object {
            "label": "One",
            "value": "one",
          },
          Object {
            "label": "Two",
            "value": "two",
          },
        ],
        "placeholder": "ojs_select_placeholder",
        "prefix": undefined,
        "suffix": undefined,
        "validators": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": undefined,
          "disabled": false,
          "id": "mock-field",
          "name": "mock-field",
          "onBlur": [MockFunction],
          "onChange": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "target": Object {
                    "name": "mock-field",
                    "value": Object {
                      "label": "One",
                      "value": "one",
                    },
                  },
                },
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          },
          "plaintext": true,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<Select /> should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Select
    field={
      Object {
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    label="field0"
    meta={
      Object {
        "error": null,
        "isValidating": false,
        "plaintext": false,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      }
    }
    multi={false}
    options={
      Array [
        Object {
          "label": "One",
          "value": "one",
        },
        Object {
          "label": "Two",
          "value": "two",
        },
      ]
    }
    placeholder="ojs_select_placeholder"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Select
        arrowRenderer={[Function]}
        autosize={true}
        backspaceRemoves={true}
        backspaceToRemoveMessage="Press backspace to remove {label}"
        className="react-select-control"
        clearAllText="Clear all"
        clearRenderer={[Function]}
        clearValueText={
          <FormComponent(FormText)
            text="ojs_select_clearValue"
          />
        }
        clearable={true}
        closeOnSelect={true}
        deleteRemoves={true}
        delimiter=","
        disabled={false}
        escapeClearsValue={true}
        filterOptions={[Function]}
        id="mock-field"
        ignoreAccents={true}
        ignoreCase={true}
        inputProps={Object {}}
        invalid={false}
        isLoading={false}
        joinValues={false}
        labelKey="label"
        matchPos="any"
        matchProp="any"
        menuBuffer={0}
        menuRenderer={[Function]}
        multi={false}
        noResultsText={
          <FormComponent(FormText)
            text="ojs_select_noresults"
          />
        }
        onBlur={[Function]}
        onBlurResetsInput={true}
        onChange={[Function]}
        onCloseResetsInput={true}
        onSelectResetsInput={true}
        openOnClick={true}
        optionComponent={[Function]}
        options={
          Array [
            Object {
              "label": "One",
              "value": "one",
            },
            Object {
              "label": "Two",
              "value": "two",
            },
          ]
        }
        pageSize={5}
        placeholder={
          <FormComponent(FormText)
            text="ojs_select_placeholder"
          />
        }
        removeSelected={true}
        required={false}
        rtl={false}
        scrollMenuIntoView={true}
        searchable={true}
        simpleValue={false}
        tabSelectsValue={true}
        trimFilter={true}
        value=""
        valueComponent={[Function]}
        valueKey="value"
      />,
      "field": Object {
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      },
      "info": undefined,
      "label": "field0",
      "meta": Object {
        "error": null,
        "isValidating": false,
        "plaintext": false,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      },
      "multi": false,
      "options": Array [
        Object {
          "label": "One",
          "value": "one",
        },
        Object {
          "label": "Two",
          "value": "two",
        },
      ],
      "placeholder": "ojs_select_placeholder",
      "prefix": undefined,
      "suffix": undefined,
      "validators": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "arrowRenderer": [Function],
        "autosize": true,
        "backspaceRemoves": true,
        "backspaceToRemoveMessage": "Press backspace to remove {label}",
        "className": "react-select-control",
        "clearAllText": "Clear all",
        "clearRenderer": [Function],
        "clearValueText": <FormComponent(FormText)
          text="ojs_select_clearValue"
        />,
        "clearable": true,
        "closeOnSelect": true,
        "deleteRemoves": true,
        "delimiter": ",",
        "disabled": false,
        "escapeClearsValue": true,
        "filterOptions": [Function],
        "id": "mock-field",
        "ignoreAccents": true,
        "ignoreCase": true,
        "inputProps": Object {},
        "invalid": false,
        "isLoading": false,
        "joinValues": false,
        "labelKey": "label",
        "matchPos": "any",
        "matchProp": "any",
        "menuBuffer": 0,
        "menuRenderer": [Function],
        "multi": false,
        "noResultsText": <FormComponent(FormText)
          text="ojs_select_noresults"
        />,
        "onBlur": [Function],
        "onBlurResetsInput": true,
        "onChange": [Function],
        "onCloseResetsInput": true,
        "onSelectResetsInput": true,
        "openOnClick": true,
        "optionComponent": [Function],
        "options": Array [
          Object {
            "label": "One",
            "value": "one",
          },
          Object {
            "label": "Two",
            "value": "two",
          },
        ],
        "pageSize": 5,
        "placeholder": <FormComponent(FormText)
          text="ojs_select_placeholder"
        />,
        "removeSelected": true,
        "required": false,
        "rtl": false,
        "scrollMenuIntoView": true,
        "searchable": true,
        "simpleValue": false,
        "tabSelectsValue": true,
        "trimFilter": true,
        "value": "",
        "valueComponent": [Function],
        "valueKey": "value",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Select
          arrowRenderer={[Function]}
          autosize={true}
          backspaceRemoves={true}
          backspaceToRemoveMessage="Press backspace to remove {label}"
          className="react-select-control"
          clearAllText="Clear all"
          clearRenderer={[Function]}
          clearValueText={
            <FormComponent(FormText)
              text="ojs_select_clearValue"
            />
          }
          clearable={true}
          closeOnSelect={true}
          deleteRemoves={true}
          delimiter=","
          disabled={false}
          escapeClearsValue={true}
          filterOptions={[Function]}
          id="mock-field"
          ignoreAccents={true}
          ignoreCase={true}
          inputProps={Object {}}
          invalid={false}
          isLoading={false}
          joinValues={false}
          labelKey="label"
          matchPos="any"
          matchProp="any"
          menuBuffer={0}
          menuRenderer={[Function]}
          multi={false}
          noResultsText={
            <FormComponent(FormText)
              text="ojs_select_noresults"
            />
          }
          onBlur={[Function]}
          onBlurResetsInput={true}
          onChange={[Function]}
          onCloseResetsInput={true}
          onSelectResetsInput={true}
          openOnClick={true}
          optionComponent={[Function]}
          options={
            Array [
              Object {
                "label": "One",
                "value": "one",
              },
              Object {
                "label": "Two",
                "value": "two",
              },
            ]
          }
          pageSize={5}
          placeholder={
            <FormComponent(FormText)
              text="ojs_select_placeholder"
            />
          }
          removeSelected={true}
          required={false}
          rtl={false}
          scrollMenuIntoView={true}
          searchable={true}
          simpleValue={false}
          tabSelectsValue={true}
          trimFilter={true}
          value=""
          valueComponent={[Function]}
          valueKey="value"
        />,
        "field": Object {
          "disabled": false,
          "id": "mock-field",
          "name": "mock-field",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "value": "",
        },
        "info": undefined,
        "label": "field0",
        "meta": Object {
          "error": null,
          "isValidating": false,
          "plaintext": false,
          "stringFormatter": [MockFunction],
          "touched": false,
          "valid": true,
        },
        "multi": false,
        "options": Array [
          Object {
            "label": "One",
            "value": "one",
          },
          Object {
            "label": "Two",
            "value": "two",
          },
        ],
        "placeholder": "ojs_select_placeholder",
        "prefix": undefined,
        "suffix": undefined,
        "validators": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "arrowRenderer": [Function],
          "autosize": true,
          "backspaceRemoves": true,
          "backspaceToRemoveMessage": "Press backspace to remove {label}",
          "className": "react-select-control",
          "clearAllText": "Clear all",
          "clearRenderer": [Function],
          "clearValueText": <FormComponent(FormText)
            text="ojs_select_clearValue"
          />,
          "clearable": true,
          "closeOnSelect": true,
          "deleteRemoves": true,
          "delimiter": ",",
          "disabled": false,
          "escapeClearsValue": true,
          "filterOptions": [Function],
          "id": "mock-field",
          "ignoreAccents": true,
          "ignoreCase": true,
          "inputProps": Object {},
          "invalid": false,
          "isLoading": false,
          "joinValues": false,
          "labelKey": "label",
          "matchPos": "any",
          "matchProp": "any",
          "menuBuffer": 0,
          "menuRenderer": [Function],
          "multi": false,
          "noResultsText": <FormComponent(FormText)
            text="ojs_select_noresults"
          />,
          "onBlur": [Function],
          "onBlurResetsInput": true,
          "onChange": [Function],
          "onCloseResetsInput": true,
          "onSelectResetsInput": true,
          "openOnClick": true,
          "optionComponent": [Function],
          "options": Array [
            Object {
              "label": "One",
              "value": "one",
            },
            Object {
              "label": "Two",
              "value": "two",
            },
          ],
          "pageSize": 5,
          "placeholder": <FormComponent(FormText)
            text="ojs_select_placeholder"
          />,
          "removeSelected": true,
          "required": false,
          "rtl": false,
          "scrollMenuIntoView": true,
          "searchable": true,
          "simpleValue": false,
          "tabSelectsValue": true,
          "trimFilter": true,
          "value": "",
          "valueComponent": [Function],
          "valueKey": "value",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
